src/node/chainstate.cpp

```diff
@@ -231,20 +232,20 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
         // for the fully validated chainstate.
         chainman.ActiveChainstate().ClearBlockIndexCandidates();
 
-        auto [init_status, init_error] = CompleteChainstateInitialization(chainman, cache_sizes, options);
-        if (init_status != ChainstateLoadStatus::SUCCESS) {
-            return {init_status, init_error};
+        auto result{CompleteChainstateInitialization(chainman, cache_sizes, options)};
+        if (!result) {
+            return result;
         }
     } else {
-        return {ChainstateLoadStatus::FAILURE, _(
-           "UTXO snapshot failed to validate. "
+        return util::Error{
+           _("UTXO snapshot failed to validate. "
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r1498934743

> Could omitting a failure value be a compile-time error if the failure type is non-void?

Nice catch. That's an interesting idea but it seems extreme because it would make it difficult to call a failure value's default constructor, and impossible to call it the class is not copyable or movable.

I think in general if you don't want a failure value to be default-constructed you should pass a failure type that doesn't have a default constructor. But we could add a check for specifically for enums. The following seems to work:

```diff
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -136,6 +136,7 @@ protected:
     static void ConstructResult(Result& result, Args&&... args)
     {
         if constexpr (Failure) {
+            static_assert(sizeof...(args) > 0 || !std::is_enum_v<F>, "Refusing to default-construct enum failure value, please specify explicit value");
             result.Info().failure.emplace(std::forward<Args>(args)...);
         } else {
             result.ConstructValue(std::forward<Args>(args)...);

```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r1499084153

> But we could add a check for specifically for enums.

Latest push adds a slightly more general check that doesn't allow any scalar failure value (int, float, enum, or pointer) to be default constructed. Could adjust this condition or make it an option if default-constructing these types of values seems useful in the future.

This check is arguably too heavy handed because it is enforcing that the `result.GetFailure()` value is not default-constructed even though it technically has a default constructor. Even without the check, the Result object would be still constructed in a failure state and `if (result)` and `if (result.has_value())` would be false.  But it seems safest to start off requiring explicit scalar failure values in case they were omitted by accident.

Reply:
Nice, the compile-time check can be tested with:
```diff
diff --git a/src/test/result_tests.cpp b/src/test/result_tests.cpp
index 261e1f0b9b..053b45fceb 100644
--- a/src/test/result_tests.cpp
+++ b/src/test/result_tests.cpp
@@ -28,0 +29,10 @@ BOOST_AUTO_TEST_SUITE(result_tests)
+enum class DummyError {
+    FAILURE_1,
+    FAILURE_2,
+    FAILURE_3,
+};
+
+util::Result<void, DummyError> VoidDummyError() {
+    return util::Error{Untranslated("void fail.")};
+}
+
@@ -203,0 +214,3 @@ BOOST_AUTO_TEST_CASE(check_returned)
+
+    auto res{VoidDummyError()};
+    BOOST_CHECK(res.GetFailure() == DummyError::FAILURE_1);
```